### PR TITLE
Add docker build doc

### PR DESCRIPTION
**Pull request by wangkuiyi**
Fixes https://github.com/PaddlePaddle/Paddle/issues/3485
Fixes https://github.com/PaddlePaddle/Paddle/issues/3690

### CODE REVIEW

File: doc/howto/dev/build_cn.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+  另外，对于习惯使用Windows和MacOS的开发者来说，使用Docker就不用配置交叉编译环境了。
+
+- 我可以选择不用Docker吗？
+
+  当然可以。大家可以用把开发工具安装进入 Docker image 一样的方式，把这些工具安装到本机。这篇文档介绍基于 Docker 的开发流程，是因为这个流程比其他方法都更简便。
+
+- 学习 Docker 有多难？
+
+  理解 Docker 并不难，大概花十分钟看一遍 https://zhuanlan.zhihu.com/p/19902938 即可。这可以帮您省掉花一小时安装和配置各种开发工具，以及切换机器时需要新安装的辛苦。别忘了 PaddlePaddle 更新可能导致需要新的开发工具。更别提简化问题复现带来的好处了。
```
Comment (luotao1) [inline doc/howto/dev/build_cn.md]: [分布式机器学习的故事：Docker改变世界](https://zhuanlan.zhihu.com/p/19902938)

File: doc/howto/dev/build_cn.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  (global-set-key "\C-cc" 'compile)
+  (setq compile-command
+   "docker run --rm -it -v $(git rev-parse --show-toplevel):/paddle paddle:dev")
+  ```
+
+  就可以按 `Ctrl-C` 和 `c` 键来启动编译了。
+
+- 可以并行编译吗？
+
+  是的。我们的 Docker image 运行一个 Bash 脚本 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/docker/build.sh 。这个脚本调用 `make -j$(nproc)` 来启动和 CPU 核一样多的进程来并行编译。
```
Comment (luotao1) [inline doc/howto/dev/build_cn.md]: [Bash 脚本](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/docker/build.sh)

File: doc/howto/dev/build_cn.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+- Docker on Windows/MacOS？
+
+  Docker 在 Windows 和 MacOS 都可以运行。不过实际上是运行在一个 Linux 虚拟机上。可能需要注意给这个虚拟机多分配一些 CPU 和内存，以保证编译高效。具体做法请参考 https://github.com/PaddlePaddle/Paddle/issues/627 。
```
Comment (luotao1) [inline doc/howto/dev/build_cn.md]: [Issue #627](https://github.com/PaddlePaddle/Paddle/issues/627)

File: doc/howto/dev/build_en.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+## General Process
+
+1. Retrieve source code.
+
+   ```bash
+   git clone https://github.com/paddlepaddle/paddle
+   ```
+
+2. Install build tools.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: would this be clear if we call this step as: Create docker image for building Paddlepaddle binary

File: doc/howto/dev/build_en.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
+# Build PaddlePaddle from Source Code and Run Unit Test
+
+## What Developers Need
+
+To contribute to PaddlePaddle, you need
+
+1. A computer -- Linux, BSD, Windows, MacOS, and
+1. Docker.
+
+Nothing else.  Not even Python and GCC, because you can install all build tools into a Docker image.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: do we need to add bit more explanation: think the docker container as a standalone factory, it takes paddlepaddle source code and produce paddlepaddle binary. Developer plays the role of a regular consumer who doesn't need to know any detail of the factory, and just use/test the product of the factory.

File: doc/howto/dev/build_en.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+   cd paddle; docker build -t paddle:dev .
+   ```
+
+3. Build from source.
+
+   ```bash
+   docker run -v $PWD:/paddle paddle:dev
+   ```
+
+   This builds a CUDA-enabled version and writes all binary outputs to directory `./build` of the local computer, other than the Docker container.  If we want to build only the CPU part, we can type
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: how about '<source>/build'

File: doc/howto/dev/build_en.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+
+## Docker, Or Not?
+
+- What is Docker?
+
+  If you haven't heard of it, consider it something like Python's virtualenv.
+
+- Docker or virtual machine?
+
+  Some people compare Docker with VMs, but Docker doesn't virtualize any hardware, and it doesn't run a guest OS.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: which means there is no compromise on performance.

File: doc/howto/dev/build_en.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+- Why Docker?
+
+  Using a Docker image of build tools standardize the building environment, and easier for others to reproduce your problem, if there is any, and help.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: Using a Docker image of build tools standardize the building environment, which makes it easier for others to reproduce your problem and easier to help.

File: doc/howto/dev/build_en.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+  Also, some build tools don't run on Windows or Mac or BSD, but Docker runs almost everywhere, so developers can use whatever computer they want.
+
+- Can I don't use Docker?
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: Can I choose not to use Docker?

File: doc/howto/dev/build_en.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+  Sure, you don't have to install build tools into a Docker image; instead, you can install them onto your local computer.  This document exists because Docker would make the development way easier.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: you can install them **in** your local computer.

File: doc/howto/dev/build_en.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+- How difficult is it to learn Docker?
+
+  It takes you ten minutes to read https://docs.docker.com/get-started/ and saves you more than one hour to install all required build tools, configure them, and upgrade them when new versions of PaddlePaddle require some new tools.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: It takes you ten minutes to read https://docs.docker.com/get-started/ and saves you more than one hour to install all required build tools, configure them, especially when new versions of PaddlePaddle require some new tools. Not even to mention the time saved when other people trying to reproduce the issue you have.

File: doc/howto/dev/build_en.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+- Docker requires sudo
+
+  An owner of a computer has the administrative privilege, a.k.a., sudo.  If you use a shared computer for development, please ask the administrator to install and configure Docker.  We will do our best to support rkt, another container technology that doesn't require sudo.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: An owner of a computer has the administrative privilege, a.k.a., sudo, and Docker requires this privilege to work properly.

File: doc/howto/dev/build_en.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+   To run all unit tests using the first GPU of a node:
+
+   ```bash
+   NV_GPU=0 nvidia-docker run -v $PWD:/paddle paddle:dev bash -c "cd /paddle/build; ctest"
+   ```
+
+   If we used `WITH_GPU=OFF` at build time, it generates only CPU-based unit tests, and we don't need nvidia-docker to run them.  We can just run
+
+   ```bash
+   docker run -v $PWD:/paddle paddle:dev bash -c "cd /paddle/build; ctest"
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: looks this document is majorly for Docker muggles, do we need to add some explanations to all the commands and parameters used above?

File: doc/howto/dev/build_en.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+  (global-set-key "\C-cc" 'compile)
+  (setq compile-command
+   "docker run --rm -it -v $(git rev-parse --show-toplevel):/paddle paddle:dev")
+  ```
+
+  so they could type `Ctrl-C` and `c` to build PaddlePaddle from source.
+
+- Does Docker do parallel building?
+
+  Our building Docker image runs a Bash script https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/scripts/docker/build.sh, which calls `make -j$(nproc)` to starts as many processes as the number of your processors.
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: to starts as many processes as the number of **your CPU cores**.

File: doc/howto/dev/build_en.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+- Docker on Windows/MacOS?
```
Comment (putcn) [inline doc/howto/dev/build_en.md]: We might want to group "Docker requires sudo" section and "Docker on Windows/Mac" in to "Some Gotchas" or "trouble shooting" section?

File: doc/howto/dev/build_en.md (hunk 6 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: The fact is that all output binary files are in the build directory under the source code root directory.

File: doc/howto/dev/build_cn.md (hunk 1 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_cn.md]: Done

File: doc/howto/dev/build_cn.md (hunk 2 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_cn.md]: Done

File: doc/howto/dev/build_cn.md (hunk 3 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_cn.md]: Done

File: doc/howto/dev/build_en.md (hunk 4 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Done

File: doc/howto/dev/build_en.md (hunk 5 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: I am afraid that it is intractable to explain Docker clearly enough in this document, so I have a hyperlink later pointing to an official Docker document.

File: doc/howto/dev/build_en.md (hunk 6 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: I changed to 

> to the sub-directory `build` in the source root directory on the local computer,

File: doc/howto/dev/build_en.md (hunk 7 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Great point! Done.

File: doc/howto/dev/build_en.md (hunk 8 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Thanks! Done.

File: doc/howto/dev/build_en.md (hunk 9 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Thanks! Done.

File: doc/howto/dev/build_en.md (hunk 10 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Done.

File: doc/howto/dev/build_en.md (hunk 11 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Great rephrase! Done.

File: doc/howto/dev/build_en.md (hunk 12 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Done!

File: doc/howto/dev/build_en.md (hunk 13 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: You are right. Done.

File: doc/howto/dev/build_en.md (hunk 14 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Done

File: doc/howto/dev/build_en.md (hunk 15 above)
Comment (wangkuiyi) [reply doc/howto/dev/build_en.md]: Great point! Done.

**Comment by putcn** [comment]
just followed the instructions to have a trial run on the CPU version, looks all command works as expected, except the process stuck at the memory unit test, I might need to adjust Docker's memory setting.

**Comment by putcn** [comment]
LGTM

**Comment by putcn** [comment]
https://github.com/PaddlePaddle/Paddle/issues/3690

**Pull request merged**